### PR TITLE
Modify Build/Command to use Path.Build.t for targets

### PR DESCRIPTION
**Pull request by rgrinberg**
* Build.declare_targets
* Build.fail
* Build.of_result
* Build.of_result_map
* Build.action
* Build.action_dyn
* Build.write_file
* Build.write_file_dyn
* Build.copy
* Build.copy_and_add_line_directive
* Build.symlink
* Build.create_tree
* Build.remove_tree
* Build.mkdir
* Build.merge_files_dyn
* Command.Target
* Command.Hidden_target
* Command.run
* Module_compilation.ocamlc_i
* Ocamlobjinfo.rules

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

### CODE REVIEW

**Comment by rgrinberg** [comment]
I took the liberty of porting  more code in Build and Build_system. Everything seems good so far but we'll definitely @aalekseyev to take another look.

File: src/build_system.ml [hunk 1]
```diff
@@ -1528,7 +1550,7 @@ end = struct
   let rules_for_files rules deps =
     Dep.Set.paths deps ~eval_pred
     |> Path.Set.fold ~init:Rule.Set.empty ~f:(fun path acc ->
-      match Path.Map.find rules path with
+      match Path.Build.Map.find rules (Path.as_in_build_dir_exn path) with
```
Comment (aalekseyev) [inline src/build_system.ml]: Maybe this shouldn't raise? I'm not sure how exactly `request` is parsed, but it seems OK (even though redundant) for the user to say that they want a source file to be up to date.

**Comment by aalekseyev** [comment]
Overall looks good to me.

**Comment by rgrinberg** [comment]
@aalekseyev addressed your comments and also ported a bit more code in Build_system to Path.Build.t

File: src/install_rules.ml [hunk 2]
```diff
@@ -272,6 +272,7 @@ let symlink_installed_artifacts_to_build_install
     let dst =
       Path.append (Path.build install_dir)
```
Comment (aalekseyev) [inline src/install_rules.ml]: Not the fault of this PR, but this `dst` definition is super-ugly. I guess `relative_installed_path` should return a `Path.Relative.t` (or `Path.Local.t` in the new terminology that we're yet to adopt).

Comment (rgrinberg) [reply src/install_rules.ml]: Indeed that is the idea.

**Pull request merged**